### PR TITLE
Refactoring scalers and adding 2 new ones

### PR DESCRIPTION
**Pull request by Nurgo**
Hi @Apaczer and @tiopex, thank you very much for the great job you've done in bringing this emulator version back to life!

Here is my modest contribution:
- Reworked the 2 existing scalers by factoring the code a bit.
- Replaced the `COLORMIX` macro with `Weight1_1`, which does the same thing in a slightly more optimized way.
- Added a new horizontal scaler that uses linear interpolation (using the optimized macros) for better results.
- Added added a new fullscreen scaler that uses true bilinear interpolation (using the optimized macros).
- Added another fullscreen scaler that uses only row and column duplication (with smoothing) for greater speed.

I've also updated the menu as follows:
- Unscaled
- Scaled => fullscreen scaling with bilinear interpolation
- Scaled (fast) => fullscreen scaling with smoothed row and column insertion
- Horizontal Scaled => horizontal scaling with linear interpolation
- Horizontal Scaled (fast) => horizontal scaling with smoothed column insertion

I let you check the code and see if it's worth integrating it into the main branch.

### CODE REVIEW

**Comment by Apaczer** [comment]
@Nurgo this look promising, especially the new bilinear interpolation for FullScreen scaling!

I run some quick benchmarks to see diff, although found hard to make those test consisent (how do you guys run your speed tests, did u run any basic bechmark❓) - only found `kmc_fpstest.sfc` ROM but it would cumbersome to do video edits. Generally there were negligible performance difference between old and new (fast) FS/HZ scaling.

About visuals:
- definetly will be replacing old_FS with new_fast_FS (more pleasing pixels placement). 👍 
- good -> FS new (bilinear) and HZ new (linear) scalers, worth adding 👍 
- couldn't spot the diff between old_HZ and new_fast_HZ, do u have any comparison pics to tell the difference? ❓ 

In comming days I'm planning to do stable release mainly inline with former version (without bogus issues), so would likely review code and merge after initial release and then do another.

**Comment by Nurgo** [comment]
Yes, it's not easy to make benchmarks. Personally, I tested a fairly heavy game running at around 35 FPS (Secreat of Mana).  The 3 fastest scalers are pretty hard to tell apart, with a difference fluctuating between 1 and 2 FPS between them. Only the FS bilinear is significantly slower (4 FPS less than the others). But I think it's worth keeping, as there are quite a few games that run at 60 FPS with it.

I don't know the `kmc_fpstest.sfc` ROM, where can I find it?

Indeed there is no visual difference between old_HZ and new_fast_HZ as they are equivalent, I just replaced `COLORMIX` by `Weight1_1` which uses fewer instructions to do the same thing, but I don't think the performance difference is visible.

**Comment by Apaczer** [comment]
> Yes, it's not easy to make benchmarks.

I ended up compiling `pidstat` from SYSSTAT pkg, to have a reliable bench tool, but left with even more unsure results by running:
```
pid=$(pidof <prog_name>)
pidstat -u -r -p $pid 1
```
also with `top`
```
while (true);do top -b -d 1| grep "$pid";done
```

> I don't know the kmc_fpstest.sfc ROM, where can I find it?

https://kidmarscat.itch.io/sfcfpstest - so u have to record flashing frames and see how long it took for that - real hw fps counter.

Either way there are small fluctuation in RAM/CPU usage from run-tu-run of same binary so hard to tell exact difference with small bench. spread.

**Comment by Nurgo** [comment]
Thanks for the link. I tried to run this ROM on my PowKiddy v90 but it doesn't seem to work. Are you doing your benchmarks on QEMU or on real hardware?

I've also noticed a difference in performance between the new open source version of DrPocketSNES and the bitrider version. For example, in Secreat of Mana, even with scaling disabled, FPS often fall below 40, whereas with the old version it was running at 60 FPS. Did I do something wrong when compiling? Or could it have something to do with the fact that I'm still running Miyoo CFW 1.3.3? Or are there still things to optimize in the new version?

**Comment by Apaczer** [comment]
> I tried to run this ROM on my PowKiddy v90 but it doesn't seem to work

Strange, it does work on my V90, athough on MiyooCFW 2.0.0.

> Did I do something wrong when compiling?

I believe not, the re-compiled version still performs worse than from bittrider's release, so we may have missed smth. Maybe by adding `-flto` ❓ . That's why I'm using PGO builds (for releases) to "turn the tables" and make final binary even faster.

However you should check if all settings are in-line in both versions, especially "Advanced hacks" cuz I disabled intentionally snesadvance.dat read by default - it may brake some games.

**Comment by Nurgo** [comment]
Wow you're right, your 7.2.2 release is almost 15 FPS faster than the version I compiled, impressive! You know your stuff 👍

File: src/miyoo_sdk.c [hunk 1]
```diff
@@ -646,94 +631,354 @@ void gp_Reset(void)
...
+			x = 64;
+			pSource += 32;
+			ScaleLineSlow(pTarget, pSource, x);
+			pSource += 256 + 32;
+			pTarget += 320;
+		}
+	}
+}
 
-			if (pos == 0)
```
Comment (Apaczer) [inline src/miyoo_sdk.c:701]: do u happen to know why there's position check for 224 source height in fullscreen mode? I noticed you've discarded that part from your scalers.

**Comment by Apaczer** [comment]
I've been running into couple of issues in[ h=239 mode](https://forums.nesdev.org/viewtopic.php?t=23870), mostly solved now but took me a few days to realize there are some settings on target (my guess its one of the "hacks" fault) issue and it would glitch a few lines in the middle of screen. Other than that, I have refactored scaling options to make it more user friendly.

> 7.2.2 release is almost 15 FPS faster than the version

I suggest you try to enable "Audio Hack" and switch to SPC700 APU, there are significant perf. boost. There's also bleeding edge `drpocketsnes_fast` build, but not many games even startup in this version.

File: src/miyoo_sdk.c (hunk 1 above)
Comment (Nurgo) [reply src/miyoo_sdk.c:701]: This corresponds to the old vertical scaling algorithm, which duplicates a line every 14 lines (calculating it twice). In the new algorithm, the vertical pattern is hard-coded, so it's no longer needed.

**Comment by Nurgo** [comment]
Looks good to me!
Well spotted the problem with `gp_video_RGB_setscaling_fast`, which was still using `ScaleLineSlow` in H239 mode 👍

**Pull request merged**